kvbc/src/categorization/block_merkle_category.cpp

```diff
@@ -0,0 +1,398 @@
+        // 0 is a tombstone sentinel
+        versions.push_back(std::nullopt);
+      } else {
+        versions.push_back(version.block_id);
+      }
+    } else if (status.IsNotFound()) {
+      versions.push_back(std::nullopt);
+    } else {
+      throw std::runtime_error{"BlockMerkleCategory multiGet() failure: " + status.ToString()};
+    }
+  }
+}
+
+void BlockMerkleCategory::multiGetLatest(const std::vector<std::string>& keys,
+                                         std::vector<std::optional<MerkleValue>>& values) const {
+  auto hashed_keys = hashedKeys(keys);
+  std::vector<std::optional<BlockId>> versions;
+  versions.reserve(keys.size());
+  multiGetLatestVersion(hashed_keys, versions);
+
+  // Generate the set of versioned keys for all keys that have latest versions
+  auto versioned_keys = std::vector<Buffer>{};
+  auto found_versions = std::vector<BlockId>{};
+  for (auto i = 0u; i < hashed_keys.size(); i++) {
+    if (versions[i]) {
+      found_versions.push_back(*versions[i]);
+      versioned_keys.push_back(serialize(VersionedKey{KeyHash{hashed_keys[i]}, *versions[i]}));
+    }
+  }
+
+  values.clear();
+  // Optimize for all keys existing (having latest versions)
+  if (versioned_keys.size() == hashed_keys.size()) {
+    return multiGet(versioned_keys, found_versions, values);
+  }
+
+  // Retrieve only the keys that have latest versions
+  auto retrieved_values = std::vector<std::optional<MerkleValue>>{};
+  retrieved_values.reserve(versioned_keys.size());
+  multiGet(versioned_keys, found_versions, retrieved_values);
+
+  // Merge any keys that didn't have latest versions along with the retrieved keys.
+  auto value_index = 0u;
+  for (auto& version : versions) {
+    if (version) {
+      values.push_back(retrieved_values[value_index]);
+      ++value_index;
+    } else {
+      values.push_back(std::nullopt);
+    }
+  }
+  ConcordAssertEQ(values.size(), keys.size());
+}
+
+void BlockMerkleCategory::putKeys(NativeWriteBatch& batch,
+                                  uint64_t block_id,
+                                  const std::vector<KeyHash>& hashed_added_keys,
+                                  const std::vector<KeyHash>& hashed_deleted_keys,
+                                  BlockMerkleInput& updates) {
+  auto kv_it = updates.kv.begin();
+  for (auto key_it = hashed_added_keys.begin(); key_it != hashed_added_keys.end(); key_it++) {
+    // Write the versioned key/value pair used for direct key lookup
+    batch.put(BLOCK_MERKLE_KEYS_CF, serialize(VersionedKey{*key_it, block_id}), kv_it->second);
+
+    // Put the latest version of the key
+    batch.put(BLOCK_MERKLE_LATEST_KEY_VERSION, key_it->value, serialize(LatestKeyVersion{block_id}));
+
+    kv_it++;
+  }
+
+  for (auto key_it = hashed_deleted_keys.begin(); key_it != hashed_deleted_keys.end(); key_it++) {
+    // Add the new tombstone to the key versions. We use 0 here to avoid the serialization overhead of a
+    // variant/optional.
```

Comment:
I was thinking whether deleting a key at some block should treat it as a new key version, but without a value. On the other hand, though, the API would look a bit weird in that `getLatest()` will return nullopt, whereas `getLatestVersion()` will return some version. That's what I did in the versioned kv category, but I think let's decide what is better and unify them.

What do you think @andrewjstone @cloudnoize ?

Reply:
I think that intuitively once a key is deleted we should return a nullopt on both cases.
And will allow redefining the key as a new key if it will be added in the future. 
If we need the information when the key was deleted maybe it's worth adding ```deletedAt``` method.
wdyt @dartdart26

Reply:
Yes, that's a good point.

I was also thinking of a case that key K is set some value at block 1 and then the key is deleted at block 5. I assume we need to give some sort of an indication to users that the key was deleted in block 5.

Reply:
If a key has been deleted, `getLatest` should always return `std::nullopt`. However, if a user asks for a key from a prior version, we return it if it exists. They likely want to see the key as they know it exists at a given block version and want to see the value for that block. The user can tell if it has been deleted by asking for the latest version if they care. `getLatestVersion` will return `nullopt` also in the case that the key has been deleted. It replaces the 0 tombstone for the user reply. 

We cannot implement a `deletedAt` method without using a std::variant to encode a tombstone. I originally had that, but it adds overhead to every latest key lookup for minimum gain. I also don't see a use case for it at the moment. I think we should only implement such a perf impacting change when we have a compelling use case.

Reply:
I updated the code in https://github.com/vmware/concord-bft/pull/1120 so that we store a tombstone with a single bit.